enhancements/network/on-prem-service-load-balancers.md

```diff
@@ -0,0 +1,405 @@
+---
+title: on-prem-service-load-balancers
+authors:
+  - "@russellb"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-05-14
+last-updated: 2020-05-14
+status: proposed
+---
+
+# Service Load Balancers for On Premise Infrastructure
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+We do not currently support [Services of
+type=LoadBalancer](https://kubernetes.io/docs/concepts/services-networking/#loadbalancer)
+(Service Load Balancers, or SLBs) for OpenShift in bare metal environments.
+While bare metal clusters are of primary interest, we hope to find a solution
+that would apply to other on-premise environments that don’t have native load
+balancer capabilities available (a cluster on VMware, RHV, or OpenStack, for
+example).
+
+## Motivation
+
+Service Load Balancers are a common way to expose applications on a cluster. We
+highly value clusters using on premise infrastructure, but do not support this
+feature in that context.  We aim to fill this gap with an optional range of
+capabilities.
+
+### Goals
```

Comment:
Do we need to define a goal for Network policy support?

Reply:
I don't think MetalLB would be involved in NetworkPolicy.  MetalLB handles IPAM and ensuring traffic gets to the right node for a given load balancer IP.  Once the traffic gets there, it's up to the network solution in use to get the traffic to the right place.  That'd be kube-proxy or whatever its equivalent is in a given setup.

Reply:
My bad, I wasn't clear. 
IIUC, MetalLB is responsible for: 
   1. IPAM & ensuring traffic gets to the right node for a given load balancer IP
   2. Adds some IPtable rules to manipulate the packet (change the external-IP) 

So, my concern was if the plugin responsible for NetworkPolicy should be aware to step '2' to be able to enforce correctly  the NetworkPolicy

Reply:
AFAIK it's only doing point 1 and not point 2.  It's up to something else to make sure the right thing happens to the traffic once it hits the host with that external IP as a destination.

Reply:
You are right -  MetalLB doesn't create any IPtables rules.
I'll create a LB service in my env and investigate the IPtables rules.
Thanks

Reply:
Note that [the NetworkPolicy docs](https://kubernetes.io/docs/concepts/services-networking/network-policies/#behavior-of-to-and-from-selectors) explicitly allow basically any behavior on cluster ingress packets:

> Cluster ingress and egress mechanisms often require rewriting the source or destination IP of packets. In cases where this happens, it is not defined whether this happens before or after NetworkPolicy processing, and the behavior may be different for different combinations of network plugin, cloud provider, Service implementation, etc.
> 
> In the case of ingress, this means that in some cases you may be able to filter incoming packets based on the actual original source IP, while in other cases, the "source IP" that the NetworkPolicy acts on may be the IP of a LoadBalancer or of the Pod's node, etc.